nixos/modules/services/continuous-integration/hercules-ci-agent/default.nix

```diff
@@ -0,0 +1,99 @@
+/*
+
+This file is for NixOS-specific options and configs.
+
+Code that is shared with nix-darwin goes in common.nix.
+
+ */
+
+{ pkgs, config, lib, ... }:
+
+let
+
+  inherit (lib) mkIf mkDefault;
+
+  cfg = config.services.hercules-ci-agent;
+
+  defaultUser = "hercules-ci-agent";
+  defaultUserDetails = {
+    name = defaultUser;
+    home = "/var/lib/hercules-ci-agent";
+    description = "System user for the Hercules Agent";
+    isSystemUser = true;
+    createHome = true;
+  };
+
+  command = "${cfg.package}/bin/hercules-ci-agent --config ${cfg.tomlFile}";
+  testCommand = "${command} --test-configuration";
+
+in
+{
+  imports = [
+    ./common.nix
+    ./deploy-keys.nix
+  ];
+
+  config = mkIf cfg.enable {
+
+    services.hercules-ci-agent.user = mkDefault defaultUser;
+    services.hercules-ci-agent.extraOptions.baseDirectory = lib.mkDefault defaultUserDetails.home;
+
+    systemd.services.hercules-ci-agent = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+      path = [ config.nix.package ];
+      serviceConfig = {
+        User = cfg.user;
+        ExecStart = command;
+        Restart = "on-failure";
+        RestartSec = 120;
+        StartLimitBurst = 30 * 1000000; # practically infinite
+      };
+    };
+
+    systemd.paths.hercules-ci-agent-restart-files = {
+      wantedBy = [ "hercules-ci-agent.service" ];
+      pathConfig = {
+        Unit = "hercules-ci-agent-restarter.service";
+        PathChanged = [ cfg.effectiveConfig.clusterJoinTokenPath ] ++ lib.optional (cfg.effectiveConfig ? binaryCachesPath) cfg.effectiveConfig.binaryCachesPath;
+      };
+    };
+
+    systemd.services.hercules-ci-agent-restarter = {
+      serviceConfig.Type = "oneshot";
+      script = ''
+        # Wait a bit, with the effect of bundling up file changes into a single
+        # run of this script and hopefully a single restart.
+        sleep 10
+        if systemctl is-active --quiet hercules-ci-agent.service; then
+          if ${testCommand}; then
+            systemctl restart hercules-ci-agent.service
+          else
+            echo 1>&2 "WARNING: Not restarting agent because config is not valid at this time."
+          fi
+        else
+          echo 1>&2 "Not restarting hercules-ci-agent despite config file update, because it is not already active."
+        fi
+      '';
+    };
```

Comment:
Can you explain a bit why this is necessary?

Reply:
Changes in the secrets do not affect the unit in any way that would case a restart, which is currently necessary to reload the secrets.

Reply:
Please add a comment to above this, so people don't need to dig up this PR, but see if next to the code.

Secret handling, and reload behaviour are somewhat intertwined with https://github.com/NixOS/rfcs/pull/59 - I'd feel much better if hercules-ci wouldn't implement its own thing here, but behaved similar to other NixOS services - which currently don't magiclaly reload on file changes (at least if the daemon itself doesn't watch these files and does so on its own).

Reply:
The reload behavior was expected by users, as shown by support requests. The nix-darwin module has the same reloading, which is trivially easy with launchd.
I'm planning to integrate reloading into the agent executable but for now the gap can be bridged by this unit.

I'll add the comment.